File: 37-sudoku-solver/sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        backtracking(board);
    }

    private boolean backtracking(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == '.') {
                    for (char num = '1'; num <= '9'; num++) {
                        if (isValid(board, i, j, num)) {
                            board[i][j] = num;
                            if (backtracking(board)) {
                                return true;
                            }
                            board[i][j] = '.'; // Undo move if it leads to a dead end
                        }
                    }
                    return false; // No valid number was found, backtrack
                }
            }
        }
        return true; // Board is complete
    }

    private boolean isValid(char[][] board, int row, int col, char num) {
        for (int i = 0; i < 9; i++) {
            // Check the row, column, and 3x3 square
            if (board[row][i] == num || board[i][col] == num ||
                    board[(row / 3) * 3 + i / 3][(col / 3) * 3 + i % 3] == num) {
                return false;
            }
        }
        return true;
    }
}

File: 37-sudoku-solver/sudoku-solver.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
function solveSudoku(board) {
    backtrack(board);
}

function backtrack(board) {
    for (let i = 0; i < 9; i++) {
        for (let j = 0; j < 9; j++) {
            if (board[i][j] === '.') {
                for (let d = 1; d <= 9; d++) {
                    const num = d.toString();
                    if (isValid(board, i, j, num)) {
                        board[i][j] = num;
                        if (backtrack(board)) return true;
                        board[i][j] = '.'; // Undo move if it leads to a dead end
                    }
                }
                return false; // No valid number was found, trigger backtracking
            }
        }
    }
    return true; // Board is complete
}

function isValid(board, row, col, num) {
    for (let i = 0; i < 9; i++) {
        // Check row and column
        if (board[row][i] === num || board[i][col] === num) return false;
        // Check the 3x3 sub-box
        let boxRow = 3 * Math.floor(row / 3) + Math.floor(i / 3);
        let boxCol = 3 * Math.floor(col / 3) + (i % 3);
        if (board[boxRow][boxCol] === num) return false;
    }
    return true;
}
